File: 2025/54. Spiral Matrix/Solution.java
class Solution {

    int[][] dir = {{0, 1}, {1, 0}, {0 , -1}, {-1, 0}};

    public List<Integer> spiralOrder(int[][] matrix) {
        List<Integer> res = new ArrayList<>();
        int i = 0, j = 0, k = 0, count = 0;
        int n = matrix.length, m = matrix[0].length;
        while (count < n * m) {
            res.add(matrix[i][j]);
            if (checkBoundary(matrix, i + dir[k % 4][0], j + dir[k % 4][1])) {
                k++;
            }
            matrix[i][j] = -101;
            i += dir[k % 4][0];
            j += dir[k % 4][1];
            count++;
        }
        return res;
    }
    
    private boolean checkBoundary(int[][] matrix, int i, int j) {
        int n = matrix.length, m = matrix[0].length;
        return (i < 0) || (j < 0) ||
                (i >= n) || (j >= m) || 
                (matrix[i][j] == -101);

    }
}

File: 2025/54. Spiral Matrix/solution.py
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        dir = [(0, 1), (1, 0), (0, -1), (-1, 0)]
        k, m, n = 0, len(matrix), len(matrix[0])
        res, visited, queue = [], set(), deque([(0,0)])
        while len(res) < m * n:
            i, j = queue.popleft()
            res.append(matrix[i][j])
            visited.add((i, j))
            x, y = i + dir[k][0], j + dir[k][1]
            if (x < 0 or y < 0 or x >= m or y >= n or (x, y) in visited):
                k = (k + 1) % 4
            x, y = i + dir[k][0], j + dir[k][1]
            queue.append((x, y))
        return res